File: mockserver/mockserver-testcontainers/src/main/java/org/mockserver/testcontainers/MockServerContainer.java
package org.mockserver.testcontainers;

import com.github.dockerjava.api.model.Capability;
import com.github.dockerjava.api.model.ExposedPort;
import com.github.dockerjava.api.model.InternetProtocol;
import org.mockserver.client.MockServerClient;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.testcontainers.containers.GenericContainer;
import org.testcontainers.containers.wait.strategy.Wait;
import org.testcontainers.utility.DockerImageName;
import org.testcontainers.utility.MountableFile;

import java.util.Map;

/**
 * An enriched Testcontainers module for MockServer.
 * <p>
 * Unlike the upstream {@code org.testcontainers:mockserver} module, this module:
 * <ul>
 *   <li>Derives the Docker image tag from the client jar version so it stays in lockstep</li>
 *   <li>Supports DNS, transparent proxy, HTTP/3, initialization JSON, and other configuration helpers</li>
 *   <li>Provides direct {@link MockServerClient} wiring via {@link #getClient()}</li>
 * </ul>
 */
public class MockServerContainer extends GenericContainer<MockServerContainer> {

    private static final Logger LOG = LoggerFactory.getLogger(MockServerContainer.class);

    /**
     * Default MockServer port (HTTP, HTTPS, SOCKS, and HTTP CONNECT are all served on a single unified port).
     */
    public static final int PORT = 1080;

    private static final String IMAGE_NAME = "mockserver/mockserver";
    private static final DockerImageName DEFAULT_IMAGE = resolveDefaultImage();

    private int serverPort = PORT;
    private MockServerClient client;

    /**
     * Creates a MockServerContainer with the default image tag derived from the client jar version.
     */
    public MockServerContainer() {
        this(DEFAULT_IMAGE);
    }

    /**
     * Creates a MockServerContainer with a custom Docker image.
     *
     * @param dockerImageName the Docker image to use (must be compatible with {@code mockserver/mockserver})
     */
    public MockServerContainer(DockerImageName dockerImageName) {
        super(dockerImageName);
        dockerImageName.assertCompatibleWith(DockerImageName.parse(IMAGE_NAME));
        addExposedPort(PORT);
        // The /mockserver/status endpoint requires a PUT request, so a simple HTTP wait
        // would fail. Use a listening-port wait strategy for robustness — note this waits
        // on ALL exposed TCP ports, which is why withServerPort() replaces (not appends) the
        // exposed port so we never wait on a port MockServer is not listening on.
        waitingFor(Wait.forListeningPort());
        // Set the default server port env var so the Docker entrypoint picks it up
        withEnv("SERVER_PORT", String.valueOf(PORT));
    }

    /**
     * Returns the mapped server port on the host.
     *
     * @return the host port mapped to the MockServer container port
     */
    public int getServerPort() {
        return getMappedPort(serverPort);
    }

    /**
     * Returns the HTTP endpoint URL for the MockServer container.
     *
     * @return the HTTP endpoint in the form {@code http://host:port}
     */
    public String getEndpoint() {
        return "http://" + getHost() + ":" + getServerPort();
    }

    /**
     * Returns the HTTPS endpoint URL for the MockServer container.
     * MockServer serves HTTP and HTTPS on the same unified port.
     *
     * @return the HTTPS endpoint in the form {@code https://host:port}
     */
    public String getSecureEndpoint() {
        return "https://" + getHost() + ":" + getServerPort();
    }

    /**
     * Returns a {@link MockServerClient} connected to this container. The client is created lazily
     * on first call and cached; it is closed automatically when the container stops.
     *
     * @return a MockServerClient connected to the running container
     */
    public synchronized MockServerClient getClient() {
        if (client == null) {
            client = new MockServerClient(getHost(), getServerPort());
        }
        return client;
    }

    /**
     * Stops the cached client (if any) and the container. The cached client is stopped gracefully,
     * which sends {@code PUT /mockserver/stop} and waits up to 10 seconds for confirmation before the
     * container itself is stopped.
     */
    @Override
    public synchronized void close() {
        if (client != null) {
            try {
                client.stop();
            } catch (Exception e) {
                LOG.debug("Exception closing cached MockServerClient", e);
            } finally {
                client = null;
            }
        }
        super.close();
    }

    /**
     * Sets a custom server port. This changes the {@code SERVER_PORT} env var and replaces the
     * exposed port so the listening-port wait strategy does not wait on a port MockServer is not
     * listening on.
     *
     * @param port the port MockServer should listen on inside the container
     * @return this container instance for chaining
     */
    public MockServerContainer withServerPort(int port) {
        this.serverPort = port;
        // Replace the exposed ports rather than appending — leaving the previous (default) port
        // exposed would make Wait.forListeningPort() block on a port MockServer never binds.
        // withExposedPorts() replaces the backing list (getExposedPorts() returns a copy, so
        // clearing it has no effect).
        withExposedPorts(port);
        withEnv("SERVER_PORT", String.valueOf(port));
        return self();
    }

    /**
     * Enables DNS resolution on the specified UDP port.
     * <p>
     * Sets {@code MOCKSERVER_DNS_ENABLED=true} and {@code MOCKSERVER_DNS_PORT} to the given port.
     * The DNS port is exposed as UDP.
     *
     * @param dnsPort the UDP port for DNS resolution
     * @return this container instance for chaining
     */
    public MockServerContainer withDnsPort(int dnsPort) {
        withEnv("MOCKSERVER_DNS_ENABLED", "true");
        withEnv("MOCKSERVER_DNS_PORT", String.valueOf(dnsPort));
        exposeUdpPort(dnsPort);
        return self();
    }

    /**
     * Enables transparent proxy mode and adds the {@code NET_ADMIN} Linux capability.
     * <p>
     * Note: iptables/redirect rules are the operator's responsibility; this helper only
     * enables the mode and grants the required capability.
     *
     * @return this container instance for chaining
     */
    public MockServerContainer withTransparentProxy() {
        withEnv("MOCKSERVER_TRANSPARENT_PROXY_ENABLED", "true");
        withCreateContainerCmdModifier(cmd ->
            cmd.getHostConfig().withCapAdd(Capability.NET_ADMIN)
        );
        return self();
    }

    /**
     * Sets multiple MockServer properties as environment variables.
     * <p>
     * Keys must be in the MockServer environment variable form (e.g. {@code MOCKSERVER_LOG_LEVEL}).
     *
     * @param properties a map of MockServer env-var keys to their values
     * @return this container instance for chaining
     */
    public MockServerContainer withProperties(Map<String, String> properties) {
        properties.forEach(this::withEnv);
        return self();
    }

    /**
     * Sets a single MockServer property as an environment variable.
     * <p>
     * The key must be in the MockServer environment variable form (e.g. {@code MOCKSERVER_LOG_LEVEL}).
     *
     * @param key   the MockServer env-var key
     * @param value the value to set
     * @return this container instance for chaining
     */
    public MockServerContainer withProperty(String key, String value) {
        withEnv(key, value);
        return self();
    }

    /**
     * Copies an initialization JSON file into the container and configures MockServer to load its
     * expectations at startup.
     * <p>
     * This is one-shot startup loading (MockServer's {@code initializationJsonPath}), not ongoing
     * persistence of expectations across restarts (which is the separate {@code persistExpectations}
     * setting). The file is copied to {@code /config/initializerJson.json} and
     * {@code MOCKSERVER_INITIALIZATION_JSON_PATH} is pointed at it.
     *
     * @param hostInitJsonPath the path on the host to the initialization JSON file
     * @return this container instance for chaining
     */
    public MockServerContainer withInitializationJson(String hostInitJsonPath) {
        withCopyFileToContainer(
            MountableFile.forHostPath(hostInitJsonPath),
            "/config/initializerJson.json"
        );
        withEnv("MOCKSERVER_INITIALIZATION_JSON_PATH", "/config/initializerJson.json");
        return self();
    }

    /**
     * Sets the MockServer log level.
     *
     * @param level the log level (e.g. "INFO", "DEBUG", "WARN", "ERROR", "TRACE")
     * @return this container instance for chaining
     */
    public MockServerContainer withLogLevel(String level) {
        withEnv("MOCKSERVER_LOG_LEVEL", level);
        return self();
    }

    /**
     * Enables experimental HTTP/3 (QUIC) support on the specified UDP port.
     * <p>
     * Sets {@code MOCKSERVER_HTTP3_PORT} and exposes the port as UDP.
     *
     * @param udpPort the UDP port for HTTP/3
     * @return this container instance for chaining
     */
    public MockServerContainer withHttp3(int udpPort) {
        withEnv("MOCKSERVER_HTTP3_PORT", String.valueOf(udpPort));
        exposeUdpPort(udpPort);
        return self();
    }

    /**
     * Exposes an additional container port over UDP. {@link #addExposedPort(Integer)} only handles
     * TCP, so UDP ports are added via a create-container command modifier.
     */
    private void exposeUdpPort(int port) {
        withCreateContainerCmdModifier(cmd -> {
            ExposedPort udpPort = new ExposedPort(port, InternetProtocol.UDP);
            ExposedPort[] existing = cmd.getExposedPorts();
            ExposedPort[] updated;
            if (existing != null) {
                updated = new ExposedPort[existing.length + 1];
                System.arraycopy(existing, 0, updated, 0, existing.length);
                updated[existing.length] = udpPort;
            } else {
                updated = new ExposedPort[]{udpPort};
            }
            cmd.withExposedPorts(updated);
        });
    }

    /**
     * Resolves the default Docker image name from the client jar's Implementation-Version so the
     * container image stays in lockstep with the client library.
     * <p>
     * When the version cannot be resolved (typically when running from an unpackaged classes
     * directory, e.g. in an IDE or on a reactor test classpath without a built jar manifest), this
     * falls back to {@code mockserver/mockserver:latest} and logs a warning. {@code :latest} is a
     * mutable tag that may not match the client on the classpath — for reproducible builds, pass an
     * explicit pinned image to {@link #MockServerContainer(DockerImageName)}.
     */
    static DockerImageName resolveDefaultImage() {
        String version = MockServerClient.class.getPackage().getImplementationVersion();
        if (version != null && !version.isEmpty()) {
            return DockerImageName.parse(IMAGE_NAME + ":mockserver-" + version);
        }
        LOG.warn("Could not resolve MockServer client version from the jar manifest; falling back to " +
            "'{}:latest'. This mutable tag may not match the client library — pass an explicit pinned " +
            "image to the MockServerContainer(DockerImageName) constructor for reproducible builds.", IMAGE_NAME);
        return DockerImageName.parse(IMAGE_NAME + ":latest");
    }
}

File: mockserver-testcontainers/php/src/MockServerContainer.php
<?php

declare(strict_types=1);

namespace MockServer\Testcontainers;

use Composer\InstalledVersions;
use Testcontainers\Container\GenericContainer;
use Testcontainers\Wait\WaitForHttp;

/**
 * A Testcontainers module for MockServer.
 *
 * Starts the official {@code mockserver/mockserver} Docker image, waits for it
 * to answer {@code PUT /mockserver/status} with HTTP 200, and returns a
 * {@see StartedMockServerContainer} exposing connection helpers plus a
 * ready-wired {@see \MockServer\MockServerClient}.
 *
 * MockServer serves HTTP, HTTPS, SOCKS, and HTTP CONNECT on a single unified
 * port (default 1080).
 *
 * @example Basic usage
 *   $container = (new MockServerContainer())->start();
 *   $client = $container->getClient();
 *   $client->when(
 *       HttpRequest::request()->method('GET')->path('/hello')
 *   )->respond(
 *       HttpResponse::response()->statusCode(200)->body('world')
 *   );
 *   // point the system under test at $container->getEndpoint()
 *   $container->stop();
 */
class MockServerContainer extends GenericContainer
{
    /**
     * Default MockServer port (HTTP, HTTPS, SOCKS, and HTTP CONNECT are all
     * served on a single unified port).
     */
    public const DEFAULT_PORT = 1080;

    /** The Docker image name on Docker Hub. */
    public const IMAGE = 'mockserver/mockserver';

    /** The Composer package name of the MockServer PHP client. */
    private const CLIENT_PACKAGE = 'mock-server/mockserver-client';

    private int $mockServerPort = self::DEFAULT_PORT;

    /**
     * @param string|null $image full Docker image reference; when null the tag
     *                           is derived from the installed MockServer client
     *                           version (see {@see defaultImage()})
     */
    public function __construct(?string $image = null)
    {
        parent::__construct($image ?? self::defaultImage());
        $this->applyMockServerDefaults();
    }

    /**
     * Overrides the port MockServer listens on inside the container. Replaces
     * the exposed port so the readiness wait targets the correct port.
     */
    public function withServerPort(int $port): static
    {
        $this->mockServerPort = $port;
        // Replace (not append) the exposed port; withExposedPorts() appends.
        $this->exposedPorts = [];
        $this->applyMockServerDefaults();

        return $this;
    }

    /**
     * Sets the MockServer log level (e.g. "INFO", "DEBUG", "WARN", "ERROR", "TRACE").
     */
    public function withLogLevel(string $level): static
    {
        $this->withEnvironment(['MOCKSERVER_LOG_LEVEL' => $level]);

        return $this;
    }

    /**
     * Sets a single MockServer property as an environment variable. The key
     * must be in MockServer env-var form (e.g. "MOCKSERVER_MAX_EXPECTATIONS").
     */
    public function withMockServerProperty(string $key, string $value): static
    {
        $this->withEnvironment([$key => $value]);

        return $this;
    }

    /**
     * Copies an initialization JSON file into the container and configures
     * MockServer to load its expectations at startup.
     */
    public function withInitializationJson(string $hostInitJsonPath): static
    {
        $containerPath = '/config/initializerJson.json';
        $this->withCopyFilesToContainer([
            ['source' => $hostInitJsonPath, 'target' => $containerPath],
        ]);
        $this->withEnvironment(['MOCKSERVER_INITIALIZATION_JSON_PATH' => $containerPath]);

        return $this;
    }

    /**
     * Starts the container and returns a {@see StartedMockServerContainer}
     * wrapper exposing MockServer-specific connection helpers.
     */
    public function start(): StartedMockServerContainer
    {
        $started = parent::start();

        return new StartedMockServerContainer($started->getId(), $started->getClient(), $this->mockServerPort);
    }

    /**
     * Resolves the default Docker image, deriving the tag from the installed
     * MockServer PHP client version so the container image stays in lockstep
     * with the client. Falls back to {@code :latest} when the version cannot be
     * resolved to a clean release version.
     */
    public static function defaultImage(): string
    {
        $version = null;
        try {
            if (InstalledVersions::isInstalled(self::CLIENT_PACKAGE)) {
                $version = InstalledVersions::getPrettyVersion(self::CLIENT_PACKAGE);
            }
        } catch (\Throwable) {
            $version = null;
        }

        return self::IMAGE . ':' . self::imageTagForVersion($version);
    }

    /**
     * Maps a MockServer client version to a Docker image tag. A clean release
     * version (e.g. "7.3.0" or "v7.3.0") yields "mockserver-7.3.0"; anything
     * else (null, a dev/branch version, a snapshot) falls back to "latest".
     */
    public static function imageTagForVersion(?string $version): string
    {
        if ($version !== null && preg_match('/^v?(\d+\.\d+\.\d+)$/', $version, $m) === 1) {
            return 'mockserver-' . $m[1];
        }

        return 'latest';
    }

    /**
     * Applies the exposed port, SERVER_PORT env var, and the PUT
     * /mockserver/status readiness wait for the current MockServer port.
     */
    private function applyMockServerDefaults(): void
    {
        $this->withExposedPorts($this->mockServerPort);
        $this->withEnvironment(['SERVER_PORT' => (string) $this->mockServerPort]);
        // /mockserver/status requires a PUT and returns 200 with the bound
        // ports once MockServer is ready.
        $this->withWait(
            (new WaitForHttp($this->mockServerPort, 60000))
                ->withMethod('PUT')
                ->withPath('/mockserver/status')
                ->withExpectedStatusCode(200)
        );
    }
}
